### PR TITLE
Moving DerivedFnCollector out of ClangPlugin

### PR DESCRIPTION
**Pull request by vaithak**
This is part of the initial refactoring required to remove direct access to ClangPlugin methods.

### CODE REVIEW

File: include/clad/Differentiator/DerivedFnCollector.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#ifndef CLAD_DIFFERENTIATOR_DERIVEDFNCOLLECTOR_H
```
Comment (vgvassilev) [inline include/clad/Differentiator/DerivedFnCollector.h:1]: Can we move that header file and its friends in `lib`?

Comment (vaithak) [reply include/clad/Differentiator/DerivedFnCollector.h:1]: Any specific reason as to why we want to move this header file into lib? I thought all header files go into include directory

Comment (vgvassilev) [reply include/clad/Differentiator/DerivedFnCollector.h:1]: These header files are "local" files that are related to the build artifact not the user interface. That is, we do not need to have them installed for clad to work...

File: tools/ClangPlugin.cpp [hunk 2]
```diff
@@ -122,7 +122,7 @@ namespace clad {
       Sema& S = m_CI.getSema();
 
       if (!m_DerivativeBuilder)
-        m_DerivativeBuilder.reset(new DerivativeBuilder(S, *this));
+        m_DerivativeBuilder.reset(new DerivativeBuilder(S, *this, m_DFC));
```
[the remark this answers is not in the document]
Comment (vaithak) [reply tools/ClangPlugin.cpp:125]: not sure if we want to apply this suggestion?

File: include/clad/Differentiator/DerivedFnCollector.h (hunk 1 above)
Comment (vaithak) [reply include/clad/Differentiator/DerivedFnCollector.h:1]: I tried it, but the issue is that this file is imported by `DerivativeBuilder.h`, which is further imported by all the Visitors.

Comment (vgvassilev) [reply include/clad/Differentiator/DerivedFnCollector.h:1]: Ok, fair enough.

File: tools/ClangPlugin.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vgvassilev) [reply tools/ClangPlugin.cpp:125]: No, that’s only available in c++14 onward.

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**